text/0000-separate-context-propagation.md

```diff
@@ -0,0 +1,193 @@
+* Separation of concerns
+  *   Remove the Tracer dependency from context propagation mechanisms.
+  *   Separate distributed context into Baggage and Correlations
+* Extensibility
+  *   Allow users to create new applications for context propagation.
+  *   For example: A/B testing, encrypted or authenticated data, and new, experimental forms of observability.
+
+## Explanation
+
+# OpenTelemetry Layered Architecture
+
+![drawing](img/context_propagation_explanation.png)
+
+OpenTelemetry is a distributed program, which requires non-local, transaction-level context in order to execute correctly. Transaction-level context can also be used to build other distributed programs, such as security, versioning, and network switching programs.
+
+To allow for this extensibility, OpenTelemetry is separated into **application layer** and a **context propagation layer**. In this architecture, multiple distributed applications - such as the observability and baggage systems provided by OpenTelemetry - simultaneously share the same underlying context propagation system in order to execute their programs.
+
+
+# Application Layer
+
+## Observability API
+
+OpenTelemetry currently contains two observability systems - Tracing and Metrics – and may be extended over time. These separate systems are bound into a unified Observability API through sharing labels – a mechanism for correlating independent observations – and through sharing propagators.
+
+**Observe(context, labels…, observations...) -> context**  
+The general form for all observability APIs is a function which takes a Context, label keys, and observations as input, and returns an updated Context.
+
+**Correlate(context, label, value, hoplimit) -> context**  
+To set the label values used by all observations in the current transaction, the Observability API provides a function which takes a context, a label key, a value, and a hoplimit, and returns an updated context. If the hoplimit is set to NO_PROPAGATION, the label will only be available to observability functions in the same process. If the hoplimit is set to UNLIMITED_PROPAGATION, it will be available to all downstream services.
+
+**GetPropagator(type) -> (inject, extract)**  
+To register with the propagation system, the Observability API provides a set of propagation functions for every propagation type. 
+
+
+## Baggage API
+
+In addition to observability, OpenTelemetry provides a simple mechanism for propagating arbitrary data, called Baggage. This allows new distributed applications to be implemented without having to create new propagators.
+
+To manage the state of a distributed application, the Baggage API provides a set of functions which read, write, and remove data.
+
+**SetBaggage(context, key, value) -> context**  
```

Comment:
for these APIs, is it just returning back the instance that was passed in? or is it actually creating a copy? 

I'm not sure if it would be efficient for a language like python to rapidly create a bunch of new, but similar, objects.

Reply:
I think this was just meant as a sort of mathematical notation, and whether this is implemented as a function modifying a thread-local it does not even get as argument or having a signature that actually has a context as parameter and return value is a language-specific concern.